Index: reddwarf/nova/db/sqlalchemy/api.py
===================================================================
--- reddwarf.orig/nova/db/sqlalchemy/api.py	2012-01-12 17:21:00.000000000 -0600
+++ reddwarf/nova/db/sqlalchemy/api.py	2012-01-12 17:27:11.000000000 -0600
@@ -2203,7 +2203,7 @@
     return iscsi_target_ref.target_num
 
 
-@require_admin_context
+@require_context
 def volume_attached(context, volume_id, instance_id, mountpoint):
     session = get_session()
     with session.begin():
@@ -2249,7 +2249,8 @@
                 filter_by(id=volume_id).\
                 update({'deleted': True,
                         'deleted_at': utils.utcnow(),
-                        'updated_at': literal_column('updated_at')})
+                        'updated_at': literal_column('updated_at'),
+                        'status': 'deleted'})
         session.query(models.ExportDevice).\
                 filter_by(volume_id=volume_id).\
                 update({'volume_id': None})
@@ -2327,7 +2328,7 @@
                    all()
 
 
-@require_admin_context
+@require_context
 def volume_get_all_by_instance(context, instance_id):
     session = get_session()
     result = session.query(models.Volume).\
@@ -2408,6 +2409,7 @@
         volume_ref = volume_get(context, volume_id, session=session)
         volume_ref.update(values)
         volume_ref.save(session=session)
+    return volume_ref
 
 
 ####################
Index: reddwarf/nova/db/sqlalchemy/models.py
===================================================================
--- reddwarf.orig/nova/db/sqlalchemy/models.py	2012-01-12 17:21:22.000000000 -0600
+++ reddwarf/nova/db/sqlalchemy/models.py	2012-01-12 17:27:15.000000000 -0600
@@ -342,6 +342,7 @@
     provider_location = Column(String(255))
     provider_auth = Column(String(255))
 
+    uuid = Column(String(64))
     volume_type_id = Column(Integer)
 
 
