Index: reddwarf/nova/utils.py
===================================================================
--- reddwarf.orig/nova/utils.py	2012-01-13 14:56:36.000000000 -0600
+++ reddwarf/nova/utils.py	2012-01-13 14:57:10.000000000 -0600
@@ -204,7 +204,7 @@
 
 def ssh_execute(ssh, cmd, process_input=None,
                 addl_env=None, check_exit_code=True):
-    LOG.debug(_('Running cmd (SSH): %s'), ' '.join(cmd))
+    LOG.debug(_('Running cmd (SSH): %s'), cmd)
     if addl_env:
         raise exception.Error(_('Environment not supported over SSH'))
 
@@ -223,6 +223,8 @@
     stdout = stdout_stream.read()
     stderr = stderr_stream.read()
     stdin_stream.close()
+    stdout_stream.close()
+    stderr_stream.close()
 
     exit_status = channel.recv_exit_status()
 
@@ -235,6 +237,7 @@
                                                   stderr=stderr,
                                                   cmd=' '.join(cmd))
 
+    channel.close()
     return (stdout, stderr)
 
 
